README.md
# EfficientNets.jl

CHANGELOG.md
# EfficientNets.jl changelog

src/load.jl
function _load!(layer::Conv; weight, bias = nothing)
    weight_perm = PermutedDimsArray(weight, (4, 3, 2, 1))
    for inds in eachindex(weight_perm)
        i, j, k, m = Tuple(inds)
        layer.weight[end-i+1, end-j+1, k, m] = weight_perm[i, j, k, m]
    end
    if !isnothing(bias)
        copyto!(layer.bias, bias)
    end
    return
end

function _load!(layer::BatchNorm; γ, β, μ, σ²)
    copyto!(layer.γ, γ)
    copyto!(layer.β, β)
    copyto!(layer.μ, μ)
    copyto!(layer.σ², σ²)
    return
end

function _load!(layer::Dense; weight, bias)
    copyto!(layer.weight, weight)
    copyto!(layer.bias, bias)
    return
end

function _load!(model::MBConv, pars, base)
    layers = isa(model.layers, SkipConnection) ? model.layers.layers : model.layers

    # Expansion phase
    expansion = layers[1]
    if !isa(expansion, typeof(identity))
        _load!(expansion[1]; weight = pars[base*"._expand_conv.weight"])
        _load!(expansion[2];
            γ=pars[base*"._bn0.weight"],
            β=pars[base*"._bn0.bias"],
            μ=pars[base*"._bn0.running_mean"],
            σ²=pars[base*"._bn0.running_var"],
        )
    end

    # Depthwise convolution phase
    depthwise = layers[2]
    _load!(depthwise[1]; weight = pars[base*"._depthwise_conv.weight"])
    _load!(depthwise[2];
        γ=pars[base*"._bn1.weight"],
        β=pars[base*"._bn1.bias"],
        μ=pars[base*"._bn1.running_mean"],
        σ²=pars[base*"._bn1.running_var"]
    )

    # Squeeze and Excitation layer
    squeeze = isa(layers[3], SkipConnection) ? layers[3].layers : layers[3]
    if !isa(squeeze, typeof(identity))
        _load!(squeeze[2];
            weight=pars[base*"._se_reduce.weight"],
            bias=pars[base*"._se_reduce.bias"],
        )
        _load!(squeeze[3];
            weight=pars[base*"._se_expand.weight"],
            bias=pars[base*"._se_expand.bias"]
        )
    end

    # Pointwise convolution phase
    point = layers[4]
    _load!(point[1]; weight = pars[base*"._project_conv.weight"])
    _load!(point[2];
        γ=pars[base*"._bn2.weight"],
        β=pars[base*"._bn2.bias"],
        μ=pars[base*"._bn2.running_mean"],
        σ²=pars[base*"._bn2.running_var"],
    )
    return
end

function _load!(model::EfficientNet{M}, path; load_top::Bool = true) where {M}
    pars = Pickle.Torch.THload(path)
    layers = model.layers

    # Stem
    stem = layers[1]
    _load!(stem[1]; weight=pars["_conv_stem.weight"])
    _load!(stem[2];
        γ=pars["_bn0.weight"],
        β=pars["_bn0.bias"],
        μ=pars["_bn0.running_mean"],
        σ²=pars["_bn0.running_var"]
    )

    # MBConv blocks
    for (i, block) in enumerate(layers[2])
        _load!(block, pars, "_blocks.$(i - 1)")
    end

    # Head
    head = layers[3]
    if !isa(head, typeof(identity))
        _load!(head[1]; weight=pars["_conv_head.weight"])
        _load!(head[2];
            γ=pars["_bn1.weight"],
            β=pars["_bn1.bias"],
            μ=pars["_bn1.running_mean"],
            σ²=pars["_bn1.running_var"]
        )
    end

    # Top
    top = layers[5]
    if !isa(top, typeof(identity)) && load_top
        _load!(top[3];
            weight=pars["_fc.weight"],
            bias=pars["_fc.bias"]
        )
    end
    return
end


src/mbconv.jl
struct MBConv{T}
    layers::T
end

@functor MBConv

function (m::MBConv)(input)
    return m.layers(input)
end

function Base.show(io::IO, m::MBConv)
    print(io, "MBConv(", m.layers, ")")
    return
end

"""
    MBConv(kernel, in => out; stride, pad, bias, skip_connection, expansion_ratio, se_ratio)

Mobile Inverted Residual Bottleneck Block.

# Arguments:
- `filter`: Size of the kernel for the depthwise conv phase.
- `in`: Number of input channels.
- `out`: Number of output channels.

# Keyword arguments
- `stride = 1`: Size of the stride for the depthwise conv phase.
- `expand_ratio = 1`: Expansion ratio defines the number of output channels `out_channels = input_channels * expansion_ratio`.
- `squeeze_ratio = 0`: Squeeze-Excitation ratio. Should be in `(0, 1]` range.
- `skip_connection = false`: Whether to use skip connection and drop connect.
- `dropout = 0`: Dropout rate. Should be in `(0, 1)` range.

# References:
1. https://arxiv.org/abs/1704.04861 (MobileNet v1)
2. https://arxiv.org/abs/1801.04381 (MobileNet v2)
3. https://arxiv.org/abs/1905.02244 (MobileNet v3)
"""
function MBConv(
    filter::NTuple{N,Integer},
    channels::Pair{<:Integer,<:Integer};
    stride=1,
    expand_ratio=1,
    squeeze_ratio=0,
    skip_connection=false,
    dropout=0,
    momentum=0.1f0,
    ϵ=1.0f-5
) where {N}

    ch_in, ch_out = channels
    ch_mid = ceil(Int, ch_in * expand_ratio)
    ch_sqz = max(1, ceil(Int, ch_in * squeeze_ratio))

    bias = false
    pad = SamePad()

    # layers
    layers = Chain(
        # Expansion phase (Inverted Bottleneck)
        expand_ratio == 1 ? identity : Chain(
            Conv((1, 1), ch_in => ch_mid; bias, pad),
            BatchNorm(ch_mid, swish; ϵ, momentum)
        ),

        # Depthwise convolution phase
        Chain(
            Conv(filter, ch_mid => ch_mid; groups=ch_mid, stride, bias, pad),
            BatchNorm(ch_mid, swish; ϵ, momentum),
        ),

        # Squeeze and Excitation layer
        !(0 < squeeze_ratio <= 1) ? identity : SkipConnection(
            Chain(
                AdaptiveMeanPool((1, 1)),
                Conv((1, 1), ch_mid => ch_sqz, swish; pad),
                Conv((1, 1), ch_sqz => ch_mid; pad),
            ),
            (mx, x) -> sigmoid.(mx) .* x
        ),

        # Pointwise convolution phase
        Chain(
            Conv((1, 1), ch_mid => ch_out; bias, pad),
            BatchNorm(ch_out; ϵ, momentum),
        ),
    )

    # Skip connection and drop connect
    if skip_connection && stride == 1 && ch_in == ch_out
        if 0 < dropout < 1
            layers = Chain(
                layers.layers...,
                Dropout(dropout),
            )
        end
        layers = SkipConnection(layers, +)
    end
    return MBConv(layers)
end

function compute_channels(channels_orig::Integer, params::ModelParams)
    multiplier = params.width_coef
    depth_divisor = params.depth_divisor

    if multiplier ≈ 1
        return channels_orig
    end

    channels = multiplier * channels_orig
    channels_new = max(
        params.depth_min,
        (floor(Int, channels + depth_divisor / 2) ÷ depth_divisor) * depth_divisor
    )
    if channels_new < 0.9 * channels
        return channels_new + depth_divisor
    else
        return channels_new
    end
end

function MBConvBlock(bparams::BlockParams, params::ModelParams)
    channels_in = compute_channels(bparams.channels[1], params)
    channels_out = compute_channels(bparams.channels[2], params)
    repeats = ceil(Int, bparams.repeats * params.depth_coef)

    return map(1:repeats) do i
        ch_in = i == 1 ? channels_in : channels_out
        ch_out = channels_out
        stride = i == 1 ? bparams.stride : 1

        return MBConv(
            bparams.filters,
            ch_in => ch_out;
            stride=stride,
            expand_ratio=bparams.expand_ratio,
            squeeze_ratio=bparams.squeeze_ratio,
            skip_connection=bparams.skip_connection,
            dropout=params.dropout,
            momentum=bparams.momentum,
            ϵ=bparams.ϵ
        )
    end
end


src/efficientnet.jl
struct EfficientNet{M,T}
    layers::T

    EfficientNet(M::Type{<:ModelName}, layers) = new{M,typeof(layers)}(layers)
end

function Flux.Functors.functor(::Type{<:EfficientNet}, x::EfficientNet{M}) where {M}
    function reconstruct_EfficientNet(xs)
        return EfficientNet(M, xs.layers)
    end
    return (; layers=x.layers), reconstruct_EfficientNet
end

function (m::EfficientNet)(input)
    return m.layers(input)
end

function Base.show(io::IO, m::EfficientNet{M}) where {M}
    print(io, "EfficientNet-", M, "(", m.layers, ")")
    return
end

stages(::EfficientNet{M}) where {M} = stages(M)
stages_channels(::EfficientNet{M}) where {M} = stages_channels(M)

function EfficientNet(
    T::Type{<:ModelName};
    classes::Integer=1000,
    channels::Integer=3,
    top::Bool=true,
    head::Bool=true,
    pretrained::Bool=false,
    adversarial::Bool=false,
    exact::Bool = true,
)

    # check if pretrained model exists
    if pretrained && channels != 3
        msg = "Pre-trained models are available only for `channels = 3`."
        throw(ArgumentError(msg))
    end
    if pretrained && classes != 1000 && top
        if exact
            msg = "Pre-trained models are available only for `classes = 1000`. If you want to use randomly generated last layer with different number of output classes, use `exact = false`."
            throw(ArgumentError(msg))
        else
            @warn "Pre-trained models are available only for `classes = 1000`. Last layer of returned model is randomly generated."
        end
    end
    # check
    if pretrained
        path = load_path(T; adversarial)
    end

    # build random model
    model = EfficientNet(T, efficientnet(T, classes, channels, top, head))

    # load pre-trained weights
    pretrained && _load!(model, path; load_top = classes == 1000)
    return model
end

function efficientnet(
    T::Type{<:ModelName},
    classes::Integer,
    channels::Integer,
    top::Bool,
    head::Bool,
)

    # model parameters
    bias = false
    pad = SamePad()
    width, depth, res, dropout = parameters(T)
    params = ModelParams((res, res), width, depth, 8, 8, dropout, top)
    ch_out = compute_channels(32, params)

    # MBConv blocks parameters
    blocks = (
        BlockParams{Float32,Int}(1, (3, 3), 32 => 16, 1, 1, 0.25, true, 0.01f0, 1.0f-3),
        BlockParams{Float32,Int}(2, (3, 3), 16 => 24, 2, 6, 0.25, true, 0.01f0, 1.0f-3),
        BlockParams{Float32,Int}(2, (5, 5), 24 => 40, 2, 6, 0.25, true, 0.01f0, 1.0f-3),
        BlockParams{Float32,Int}(3, (3, 3), 40 => 80, 2, 6, 0.25, true, 0.01f0, 1.0f-3),
        BlockParams{Float32,Int}(3, (5, 5), 80 => 112, 1, 6, 0.25, true, 0.01f0, 1.0f-3),
        BlockParams{Float32,Int}(4, (5, 5), 112 => 192, 2, 6, 0.25, true, 0.01f0, 1.0f-3),
        BlockParams{Float32,Int}(1, (3, 3), 192 => 320, 1, 6, 0.25, true, 0.01f0, 1.0f-3),
    )

    # head parameters
    ch_head_in = compute_channels(blocks[end].channels[2], params)
    ch_head_out = compute_channels(1280, params)

    # layers
    return Chain(
        # Stem
        Chain(
            Conv((3, 3), channels => ch_out; stride=2, bias, pad),
            BatchNorm(ch_out, swish),
        ),

        # MBConv blocks
        Chain(
            MBConvBlock(blocks[1], params)...,
            MBConvBlock(blocks[2], params)...,
            MBConvBlock(blocks[3], params)...,
            MBConvBlock(blocks[4], params)...,
            MBConvBlock(blocks[5], params)...,
            MBConvBlock(blocks[6], params)...,
            MBConvBlock(blocks[7], params)...,
        ),

        # Head
        !head ? identity : Chain(
            Conv((1, 1), ch_head_in => ch_head_out; bias, pad),
            BatchNorm(ch_head_out, swish),
        ),

        # Final linear layer
        AdaptiveMeanPool((1, 1)),
        !top ? identity : Chain(
            Flux.flatten,
            Dropout(dropout),
            Dense(ch_head_out, classes),
        ),
    )
end

function extract_features(m::EfficientNet, x)
    ids = stages(m)
    layers = (
        m.layers[1],
        m.layers[2][1:ids[1]],
        m.layers[2][(ids[1] + 1):ids[2]],
        m.layers[2][(ids[2] + 1):ids[3]],
        m.layers[2][(ids[3] + 1):ids[4]],
    )
    return Flux.extraChain(layers, x)
end


src/EfficientNets.jl
module EfficientNets

using DataDeps
using Flux
using Pickle

using Flux: @functor
using Flux: flatten
using Flux: SkipConnection

export EfficientNet
export MBConv
export B0, B1, B2, B3, B4, B5, B6, B7, B8, L2

export extract_features
export stages
export stages_channels

# Custom types
struct ModelParams{T<:Real,I<:Integer}
    image_size::NTuple{2,I}
    width_coef::T
    depth_coef::T
    depth_divisor::I
    depth_min::I
    dropout::T
    top::Bool
end

struct BlockParams{T<:Real,I<:Integer}
    repeats::I
    filters::NTuple{2,I}
    channels::Pair{I,I}
    stride::I
    expand_ratio::I
    squeeze_ratio::T
    skip_connection::Bool
    momentum::T
    ϵ::T
end

include("models.jl")
include("mbconv.jl")
include("efficientnet.jl")
include("load.jl")

function __init__()
    for T in (B0, B1, B2, B3, B4, B5, B6, B7, B8, L2)
        urls = [[url(T), urlhash(T)], [url_adv(T), urlhash_adv(T)]]
        filter!(url -> !isnothing(url[1]), urls)

        if !isempty(urls)
            register(DataDep(
                "EfficientNet-$(name(T))",
                "Pre-trained weights for EfficientNet-$(name(T)) model.",
                first.(urls),
                last.(urls),
            ))
        end
    end
end
end # module


src/models.jl
abstract type ModelName end

name(T::Type{<:ModelName}) = string(T.name.name)

"""
    parameters(::Type{<:ModelName})

Specifies parameters of the model based on the model name. The function returns a tuple of `(width, depth, resolution, dropout)`.

"""
function parameters() end

url(::Type{<:ModelName}) = nothing
urlhash(::Type{<:ModelName}) = nothing

url_adv(::Type{<:ModelName}) = nothing
urlhash_adv(::Type{<:ModelName}) = nothing

function load_path(T::Type{<:ModelName}; adversarial::Bool=false)
    file = adversarial ? url_adv(T) : url(T)
    if isnothing(file)
        type = adversarial ? "adversarial" : ""
        msg = "Pre-trained $(type) model is not available for EfficientNet-$(T)"
        throw(ArgumentError(msg))
    end
    modelpath = "EfficientNet-$(name(T))/$(basename(file))"
    return @datadep_str modelpath
end

# Predefined models
const URL_BASE = "https://github.com/lukemelas/EfficientNet-PyTorch/releases/download/1.0"

struct B0 <: ModelName end

parameters(::Type{B0}) = (1.0, 1.0, 224, 0.2)
stages(::Type{B0}) = (3, 5, 9, 16)
stages_channels(::Type{B0}) = (32, 24, 40, 112, 320)

url(::Type{B0}) = joinpath(URL_BASE, "efficientnet-b0-355c32eb.pth")
urlhash(::Type{B0}) = "355c32ebf5c7b2e19ed8f926daea5b6d71c8744a49aab3148fc90775fbaac5a3"

url_adv(::Type{B0}) = joinpath(URL_BASE, "adv-efficientnet-b0-b64d5a18.pth")
urlhash_adv(::Type{B0}) = "b64d5a18952a215048ee8541f46f0d0fb253eee749f780d22c6aa57561de5c3e"

struct B1 <: ModelName end

parameters(::Type{B1}) = (1.0, 1.1, 240, 0.2)
stages(::Type{B1}) = (5, 8, 16, 23)
stages_channels(::Type{B1}) = (32, 24, 40, 112, 320)

url(::Type{B1}) = joinpath(URL_BASE, "efficientnet-b1-f1951068.pth")
urlhash(::Type{B1}) = "f19510681ee65135bad30718410bff1f9c51452b858a6db2f61b7707e18b933d"

url_adv(::Type{B1}) = joinpath(URL_BASE, "adv-efficientnet-b1-0f3ce85a.pth")
urlhash_adv(::Type{B1}) = "0f3ce85a0ecbc16831f7f9e6202f35dd34a6642ba571984bf5281b6aa860d47a"

struct B2 <: ModelName end

parameters(::Type{B2}) = (1.1, 1.2, 260, 0.3)
stages(::Type{B2}) = (5, 8, 16, 23)
stages_channels(::Type{B2}) = (32, 24, 48, 120, 352)

url(::Type{B2}) = joinpath(URL_BASE, "efficientnet-b2-8bb594d6.pth")
urlhash(::Type{B2}) = "8bb594d63dcd2315ffd3fe5b6dc4380c3cc8ce4209bdf63a5d76963ce36845c0"

url_adv(::Type{B2}) = joinpath(URL_BASE, "adv-efficientnet-b2-6e9d97e5.pth")
urlhash_adv(::Type{B2}) = "6e9d97e5ee02bdfd722a128c4db18110999316af5e40a6f30b074ad0db1057db"

struct B3 <: ModelName end

parameters(::Type{B3}) = (1.2, 1.4, 300, 0.3)
stages(::Type{B3}) = (5, 8, 18, 26)
stages_channels(::Type{B3}) = (40, 32, 48, 136, 384)

url(::Type{B3}) = joinpath(URL_BASE, "efficientnet-b3-5fb5a3c3.pth")
urlhash(::Type{B3}) = "5fb5a3c3f1d45062d3d3dbe770f1a7daf78eed140f1c092cceef9ab79677d08e"

url_adv(::Type{B3}) = joinpath(URL_BASE, "adv-efficientnet-b3-cdd7c0f4.pth")
urlhash_adv(::Type{B3}) = "cdd7c0f4d1e8d97387d00bc8d0ccfe55364c232e9cb04ae6f1b00d20912feb33"

struct B4 <: ModelName end

parameters(::Type{B4}) = (1.4, 1.8, 380, 0.4)
stages(::Type{B4}) = (6, 10, 22, 32)
stages_channels(::Type{B4}) = (48, 32, 56, 160, 448)

url(::Type{B4}) = joinpath(URL_BASE, "efficientnet-b4-6ed6700e.pth")
urlhash(::Type{B4}) = "6ed6700e361e3878dc53f285f3df4abf9f39f6da23545cac065b7dc9467a8959"

url_adv(::Type{B4}) = joinpath(URL_BASE, "adv-efficientnet-b4-44fb3a87.pth")
urlhash_adv(::Type{B4}) = "44fb3a87ef763da422a700d9d1f62328263ce3d86074e92766d28e3e1ceb2408"

struct B5 <: ModelName end

parameters(::Type{B5}) = (1.6, 2.2, 456, 0.4)
stages(::Type{B5}) = (8, 13, 27, 39)
stages_channels(::Type{B5}) = (48, 40, 64, 176, 512)

url(::Type{B5}) = joinpath(URL_BASE, "efficientnet-b5-b6417697.pth")
urlhash(::Type{B5}) = "b641769706a245cea68d309c3d35a864b6317e686e315fe79ed3d81e78c6f51c"

url_adv(::Type{B5}) = joinpath(URL_BASE, "adv-efficientnet-b5-86493f6b.pth")
urlhash_adv(::Type{B5}) = "86493f6be52121a1908efaa228e1ea396d28d5155468b4cfbc145d8f71cede2c"

struct B6 <: ModelName end

parameters(::Type{B6}) = (1.8, 2.6, 528, 0.5)
stages(::Type{B6}) = (9, 15, 31, 45)
stages_channels(::Type{B6}) = (56, 40, 72, 200, 576)

url(::Type{B6}) = joinpath(URL_BASE, "efficientnet-b6-c76e70fd.pth")
urlhash(::Type{B6}) = "c76e70fd23f007dd6b09c8464e4e184a81d4d3d715ee42c2c89650c0be701505"

url_adv(::Type{B6}) = joinpath(URL_BASE, "adv-efficientnet-b6-ac80338e.pth")
urlhash_adv(::Type{B6}) = "ac80338ed82705fa2710279c583263071ccd249ab856f682f1be52d6da14ff4e"

struct B7 <: ModelName end

parameters(::Type{B7}) = (2.0, 3.1, 600, 0.5)
stages(::Type{B7}) = (11, 18, 38, 55)
stages_channels(::Type{B7}) = (64, 48, 80, 224, 640)

url(::Type{B7}) = joinpath(URL_BASE, "efficientnet-b7-dcc49843.pth")
urlhash(::Type{B7}) = "dcc49843b8ec5a097a8cb166eded1d7a09291093eff7735f603bcfcafce12c8e"

url_adv(::Type{B7}) = joinpath(URL_BASE, "adv-efficientnet-b7-4652b6dd.pth")
urlhash_adv(::Type{B7}) = "4652b6dd826ae720eb0ad60b03a6893dd18c18b5527781660292bd9dd75e9ccd"

struct B8 <: ModelName end

parameters(::Type{B8}) = (2.2, 3.6, 672, 0.5)

url_adv(::Type{B8}) = joinpath(URL_BASE, "adv-efficientnet-b8-22a8fe65.pth")
urlhash_adv(::Type{B8}) = "22a8fe65d42c4dc0599dfdf0519f823f7eb301ae65aa20fa32b32319767e93c8"

struct L2 <: ModelName end

parameters(::Type{L2}) = (4.3, 5.3, 800, 0.5)
